src/Locks.ts

```diff
@@ -0,0 +1,234 @@
+  InvalidArgumentError,
+  inspect,
+  type EventKey,
+  type EventListener,
+} from './utilities/EventEmitter.js';
+
+export class LockAttributes extends Map<string, string> {
+  toJSON() {
+    return Object.fromEntries(this);
+  }
+}
+
+interface LockOptions {
+  attributes: LockAttributes;
+}
+
+type LockEventMap = {
+  update: Lock;
+};
+
+export default class Locks extends EventEmitter<LockEventMap> {
+  constructor(
+    private space: Space,
+    private presenceUpdate: (update: PresenceMember['data'], extras?: any) => Promise<void>,
+  ) {
+    super();
+  }
+
+  get(id: string): Lock | undefined {
+    for (const member of this.space.members.getAll()) {
+      const request = member.locks.get(id);
+      if (request && request.status === 'locked') {
+        return { request, member };
+      }
+    }
+  }
+
+  async acquire(id: string, opts?: LockOptions): Promise<LockRequest> {
+    const self = this.space.members.getSelf();
+    if (!self) {
+      throw new Error('Must enter a space before acquiring a lock');
+    }
+
+    // check there isn't an existing PENDING or LOCKED request for the current
+    // member, since we do not support nested locks
+    let req = self.locks.get(id);
+    if (req && req.status !== 'unlocked') {
+      throw ERR_LOCK_REQUEST_EXISTS;
+    }
+
+    // initialise a new PENDING request
+    req = {
+      id,
+      status: 'pending',
+      timestamp: Date.now(),
+    };
+    if (opts) {
+      req.attributes = opts.attributes;
+    }
+    self.locks.set(id, req);
+
+    // reflect the change in the member's presence data
+    await this.updatePresence(self);
+
+    return req;
+  }
+
+  async release(id: string): Promise<void> {
+    const self = this.space.members.getSelf();
+    if (!self) {
+      throw new Error('Must enter a space before acquiring a lock');
+    }
+
+    self.locks.delete(id);
+
+    await this.updatePresence(self);
+  }
+
+  subscribe<K extends EventKey<LockEventMap>>(
+    listenerOrEvents?: K | K[] | EventListener<LockEventMap[K]>,
+    listener?: EventListener<LockEventMap[K]>,
+  ) {
+    try {
+      super.on(listenerOrEvents, listener);
+    } catch (e: unknown) {
+      if (e instanceof InvalidArgumentError) {
+        throw new InvalidArgumentError(
+          'Locks.subscribe(): Invalid arguments: ' + inspect([listenerOrEvents, listener]),
+        );
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  unsubscribe<K extends EventKey<LockEventMap>>(
+    listenerOrEvents?: K | K[] | EventListener<LockEventMap[K]>,
+    listener?: EventListener<LockEventMap[K]>,
+  ) {
+    try {
+      super.off(listenerOrEvents, listener);
+    } catch (e: unknown) {
+      if (e instanceof InvalidArgumentError) {
+        throw new InvalidArgumentError(
+          'Locks.unsubscribe(): Invalid arguments: ' + inspect([listenerOrEvents, listener]),
+        );
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  processPresenceMessage(message: Types.PresenceMessage) {
```

Comment:
Shouldn't this be private?  Not sure on support for this, but I thought the recommended approach is `#` prefix?

Reply:
It's called by the `Space` class [here](https://github.com/ably-labs/spaces/blob/main/src/Space.ts#L100), so it can't be private no.